File: 1212/main.py
'''
Copyright 2021 Christian Kramer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
'''


import os
from collections import defaultdict


def read_input():
    lines = []
    coords = defaultdict(list)
    with open(os.path.abspath("./1212/input")) as file: 
        lines = file.readlines()
        for line in lines:
            splitted = line.strip().split("-")
            coords[splitted[0]] += [splitted[1]]
            coords[splitted[1]] += [splitted[0]]
    return coords


def iterate(coords, way, seen, node, paths, twice):  
    way.append(node)
    if node == "end":
        paths.append(way)
        
    for c in coords[node]:
        if c.islower():
            if c == "start":
                continue
            if c not in seen:
                seen.append(c)
                iterate(coords, way.copy(), seen, c, paths, twice)
            elif twice and c != "end":
                seen.append(c)
                iterate(coords, way.copy(), seen, c, paths, False)
            else:
                continue
        else:
            iterate(coords, way.copy(), seen, c, paths, twice)            
        if c.islower():
            seen.pop()
    
        

def find_ways(coords):
    way = list()
    seen = list()
    paths = list()
    iterate(coords, way, seen, "start", paths, True)
    return paths

if __name__ == '__main__':
    coords = read_input()
    print(len(find_ways(coords)))

File: 1212/main/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::env;
use std::collections::HashMap;
use regex::Regex;



fn read_input(args: &Vec<String>) -> HashMap<String, Vec<String>>{
    let filename = &args[1];
    println!("filename {}", filename);
    let file = File::open(filename).unwrap();
    let reader = BufReader::new(file);
    let mut coords : HashMap<String, Vec<String>> = HashMap::new();


    for line in reader.lines() {
        let line_content = line.unwrap();
        let pairs : Vec<&str> = line_content.split("-").collect();

        if coords.contains_key(pairs[0]) {
            let vec = coords.get_mut(pairs[0]).unwrap();
            vec.push(pairs[1].to_owned());
        } else {
            let mut vec : Vec<String> = Vec::new();
            vec.push(pairs[1].to_owned());
            coords.insert(pairs[0].to_owned(), vec.to_owned());
        }
    }

    println!("{:?}", coords);
    return coords;
}

fn iter(coords: &HashMap<String, Vec<String>>, node: &String, call_node: &String, mut seen: Vec<String>, mut way: Vec<String>) {
    let re = Regex::new(r"^[A-Z]$").unwrap();


    if !re.is_match(node) && node != "start" {
        if !seen.contains(node) {
            seen.push(node.clone());
        } else {
            return;
        }
    }

    way.push(node.clone());

    if node == "end" {
        println!("end {:?}", way);
        way.pop();
    }

    println!("way {:?} seen {:?}", way, seen);

    if coords.contains_key(node) {
        for a in coords.get(node).unwrap() {
            //if !seen.contains(a) {
                iter(coords, a, node, seen.clone(), way.clone());
        
            //} else {
                // iter(coords, node, node, seen.clone(), way.clone());
            //}
        }
    } else {
        if !seen.contains(call_node){
            iter(coords, call_node, node, seen.clone(), way.clone());
        }
    }
}

fn find_ways(coords: HashMap<String, Vec<String>>) {
    let seen : Vec<String> = Vec::new();
    let way: Vec<String> = Vec::new();
    iter(&coords, &String::from("start"), &String::from(""), seen, way);
}


fn main() {
    let args: Vec<String> = env::args().collect();
    let coords = read_input(&args);
    find_ways(coords);
    
}